src/functions/projectConfig.ts

```diff
@@ -64,7 +109,34 @@ function validateSingle(config: FunctionConfig): ValidatedSingle {
     validatePrefix(config.prefix);
   }
 
-  return { ...config, source: config.source, codebase: config.codebase };
+  // Narrow to validated shapes
+  const { codebase, prefix, ignore, configDir, predeploy, postdeploy } = config;
+  if (source) {
+    const validated: ValidatedLocalSingle = {
+      source,
+      codebase,
+      ...(runtime ? { runtime } : {}),
```

Comment:
Why is this not just "runtime"? because you don't want to insert undefineds?

Reply:
ah this was because i didn't want keys to be checked during unit testing..